// drizzle-orm/pg-core/db.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PgDatabase = void 0;
const query_builders_1 = require("./query-builders");
class PgDatabase {
    constructor(
    /** @internal */
    dialect, 
    /** @internal */
    session) {
        this.dialect = dialect;
        this.session = session;
    }
    with(...queries) {
        const self = this;
        function select(fields) {
            return new query_builders_1.PgSelectBuilder(fields !== null && fields !== void 0 ? fields : undefined, self.session, self.dialect, queries);
        }
        return { select };
    }
    select(fields) {
        return new query_builders_1.PgSelectBuilder(fields !== null && fields !== void 0 ? fields : undefined, this.session, this.dialect);
    }
    update(table) {
        return new query_builders_1.PgUpdateBuilder(table, this.session, this.dialect);
    }
    insert(table) {
        return new query_builders_1.PgInsertBuilder(table, this.session, this.dialect);
    }
    delete(table) {
        return new query_builders_1.PgDelete(table, this.session, this.dialect);
    }
    execute(query) {
        return this.session.execute(query.getSQL());
    }
}
exports.PgDatabase = PgDatabase;
