// drizzle-orm/pg-core/columns/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PgColumn = exports.PgColumnBuilder = void 0;
const column_1 = require("../../column");
const column_builder_1 = require("../../column-builder");
const foreign_keys_1 = require("../foreign-keys");
class PgColumnBuilder extends column_builder_1.ColumnBuilder {
    constructor(name) {
        super(name);
        this.foreignKeyConfigs = [];
    }
    notNull() {
        return super.notNull();
    }
    default(value) {
        return super.default(value);
    }
    primaryKey() {
        return super.primaryKey();
    }
    references(ref, actions = {}) {
        this.foreignKeyConfigs.push({ ref, actions });
        return this;
    }
    /** @internal */
    buildForeignKeys(column, table) {
        return this.foreignKeyConfigs.map(({ ref, actions }) => {
            return ((ref, actions) => {
                const builder = new foreign_keys_1.ForeignKeyBuilder(() => {
                    const foreignColumn = ref();
                    return { columns: [column], foreignColumns: [foreignColumn] };
                });
                if (actions.onUpdate) {
                    builder.onUpdate(actions.onUpdate);
                }
                if (actions.onDelete) {
                    builder.onDelete(actions.onDelete);
                }
                return builder.build(table);
            })(ref, actions);
        });
    }
}
exports.PgColumnBuilder = PgColumnBuilder;
// To understand how to use `PgColumn` and `AnyPgColumn`, see `Column` and `AnyColumn` documentation.
class PgColumn extends column_1.Column {
    constructor(table, config) {
        super(table, config);
        this.table = table;
    }
}
exports.PgColumn = PgColumn;
